include/valijson/validation_results.hpp

```diff
@@ -51,6 +51,36 @@ class ValidationResults
     };
 
     /**
+     * @brief  Iterator for all validation errors.
+     *
+     * This struct is used to get a reference to an iterator for errors.
+     */
+    struct ErrorRange
+    {
+        typedef std::deque<Error>::const_iterator ErrorIterator;
+        ErrorRange(const std::deque<Error> &errors) : m_errors(errors) {};
+
+        ErrorIterator begin() {
+            return m_errors.begin();
+        }
+
+        ErrorIterator end() {
+            return m_errors.end();
+        }
+
+    private:
+        const std::deque<Error> &m_errors;
```

Comment:
@krzysztofwos, I quite like the idea of moving begin() and end() back into the ValidationResults class rather than introducing a new struct. The example you've provided shows that it leads to a very clean API, and it does make it easier to provide a guarantee about when the iterators may/may not be invalidated.

@iterion, this change would work for your use case, right?

Making pushError and popError private also seems like a good idea, but that is something I would like to tackle separately. The original intention behind the name 'ValidationResults' was that it may one day act as a container for warnings and user-generated messages as well as errors. I think this should be revisited before breaking the API.

As for the virtual destructor, I've removed it entirely in commit 3a9f13c.